### PR TITLE
improve custom template handling

### PR DESCRIPTION
**Pull request by ironashram**
The current way of loading a custom haproxy template is broken
This change will load the custom template from `custom_template/haproxy.tmpl` if template is set to true in the values file

### CODE REVIEW

**Comment by jcmoraisjr** [comment]
Hi, I could find any problem in the current implementation. How are you customizing the template (what you configure and where) and what were you expecting as a result?

**Comment by jcmoraisjr** [comment]
Hi, I could find any problem in the current implementation. How are you customizing the template (what you configure and where) and what were you expecting as a result?

**Comment by ironashram** [comment]
Hi, for me neither of the following
`{{ .Values.controller.template | indent 4 }}`
`{{ .Files.Get "haproxy.tmpl" | indent 4 }}`
results in helm3 rendering the chart successfully, i tried both adding the modified template to the values or loading the modified `haproxy.tmpl` but it looks like helm tries to interpret the the template this way hence the reason why i propose this change

As for the actual changes to the haproxy template i do the following

1. Add dummy backends for sticky tables, which i use for custom rate limiting requests/ddos protection
2. Change the position where the custom frontend snippets are being added, i do this because the custom snippet i need contains a few tcp-request/connection directives which with the default template end up being placed after some other http related directives, thus haproxy keeps throwing warnings about the wrong order on every reload

**Comment by jcmoraisjr** [comment]
Hi, regarding the template I can successfully change it with these two changes:

- remove the two spaces indentation from the first line - yaml parser doesn't like it because it'd instead interpret that the whole file would have those extra spaces, which it isn't true;
- remove the `-` from the last line, using `{{ end }}` instead of `{{- end }}` - haproxy parser doesn't like this last line without a line break and the template is removing the trailing line breaks because of the minus signal in the `haproxy.tmpl: |-` helm template.

I think that these two changes would be benefit for the functionality:

- remove the spaces in the first line of the template (controller repo)
- change from `|-` to simply `|` in the helm template

Other than that, regarding the issues you're trying to fix:

- `1.` can be configured with `config-sections`, doc [here](https://haproxy-ingress.github.io/docs/configuration/keys/#configuration-snippet)
- `2.` please provide a reproducer in the controller repo so this won't be forgotten

**Comment by ironashram** [comment]
Hi, thank you for all the details, i'm going to do a few tests with helm and the insight you gave me

Regarding the suggestions

1. I know there is this possibility from v0.13, unfortunately i can't use that yet on production due to having an old k8s version on some clusters
2. Opened --> https://github.com/jcmoraisjr/haproxy-ingress/issues/909

**Comment by ironashram** [comment]
@jcmoraisjr i made some tests and i agree with your proposed solution

1. remove the spaces in the first line of the template (controller repo)
2. change from `|-` to simply `|` in the helm template

However i'd still change `Values.template` to boolean and remove `{{ .Values.controller.template | indent 4 }}` or not make it the default behavior since it's very messy to use

**Comment by jcmoraisjr** [comment]
This cannot be changed due to backward compatibility, but there isn't any problem to create another key with a distinct behavior, maybe a `templateFile` which receives a path? How does it sound?

**Comment by ironashram** [comment]
> This cannot be changed due to backward compatibility, but there isn't any problem to create another key with a distinct behavior, maybe a `templateFile` which receives a path? How does it sound?

pushed a commit that includes this new key and keeps backwards compatibility, let me know if that works for you

File: haproxy-ingress/README.md [hunk 1]
```diff
@@ -109,6 +109,7 @@ Parameter | Description | Default
 `controller.extraContainers` | extra containers that to the haproxy-ingress-controller | `[]`
 `controller.initContainers` | extra containers that can initialize the haproxy-ingress-controller | `[]`
 `controller.template` | custom template for haproxy-ingress-controller | `{}`
+`controller.templateFile` | custom haproxy template path for haproxy-ingress-controller | `{}`
```
Comment (jcmoraisjr) [inline haproxy-ingress/README.md:112]: `""` instead in the last column

File: haproxy-ingress/templates/controller-template.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-{{- if .Values.controller.template -}}
+{{- if or .Values.controller.template .Values.controller.templateFile -}}
```
Comment (jcmoraisjr) [inline haproxy-ingress/templates/controller-template.yaml:1]: missing this same logic in the `_podtemplate.yaml`, look for `controller.template`

File: haproxy-ingress/templates/controller-template.yaml [hunk 3]
```diff
@@ -7,7 +7,10 @@ metadata:
...
   namespace: {{ .Release.Namespace }}
 data:
-  haproxy.tmpl: |-
+  haproxy.tmpl: |
+{{- if .Values.controller.templateFile -}}
+{{ .Files.Get .Values.controller.templateFile | indent 4 }}
+{{- else }}
 {{ .Values.controller.template | indent 4 }}
-{{/* .Files.Get "haproxy.tmpl" | indent 4 */}}
+{{- end }}
```
Comment (jcmoraisjr) [inline haproxy-ingress/templates/controller-template.yaml:15]: these trimleft configs broke the yaml parser, here is a config suggestion that worked for me:

```
  haproxy.tmpl: |
{{- if .Values.controller.templateFile }}
    {{- .Files.Get .Values.controller.templateFile | nindent 4 }}
{{- else }}
    {{- .Values.controller.template | nindent 4 }}
{{- end }}
```

**Review by jcmoraisjr — COMMENTED** [review]
thanks! here are a few update suggestions.

File: haproxy-ingress/templates/controller-template.yaml (hunk 2 above)
Comment (ironashram) [reply haproxy-ingress/templates/controller-template.yaml:1]: indeed, added

File: haproxy-ingress/templates/controller-template.yaml (hunk 3 above)
Comment (ironashram) [reply haproxy-ingress/templates/controller-template.yaml:15]: fixed

**Comment by jcmoraisjr** [comment]
Lgtm thanks! Merging now.

**Pull request merged**